src/vsr/journal.zig

```diff
@@ -669,117 +812,541 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+                    self.dirty.set(slot);
+                    self.faulty.set(slot);
+                },
+            }
+
+            switch (case.decision) {
+                .eql, .nil => {
+                    log.debug("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
+                .fix, .vsr => {
+                    log.warn("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
             }
 
-            // We must release before we call `recover_headers()` in case Storage is synchronous.
-            // Otherwise, we would run out of messages and reads.
             replica.message_bus.unref(read.message);
             self.reads.release(read);
+            self.recover_prepares(Slot{ .index = slot.index + 1 });
+        }
+
+        fn recover_done(self: *Self) void {
+            const replica = @fieldParentPtr(Replica, "journal", self);
+
+            log.debug("{}: recover_done: dirty={} faulty={}", .{
+                self.replica,
+                self.dirty.count,
+                self.faulty.count,
+            });
+
+            assert(self.faulty.count == 0 or replica.replica_count != 1);
+            assert(self.faulty.count >= self.dirty.count);
+            assert(self.faulty.count < slot_count);
+            if (self.headers[0].op == 0 and self.headers[0].command == .prepare) {
+                assert(self.headers[0].checksum == Header.root_prepare(replica.cluster).checksum);
+            }
 
-            self.recover_headers(offset + buffer.len, version);
+            {
+                var min_op: u64 = std.math.maxInt(u64);
+                var max_op: u64 = 0;
+                for (self.headers) |*header, slot| {
+                    if (header.command == .reserved) {
+                        assert(header.cluster == replica.cluster);
+                        assert(header.op == slot);
+                        continue;
+                    }
+
+                    assert(header.command == .prepare);
+                    assert(header.op % slot_count == slot);
+                    assert(!self.dirty.bit(Slot{ .index = slot }));
+                    assert(!self.faulty.bit(Slot{ .index = slot }));
+
+                    if (min_op > header.op) min_op = header.op;
+                    if (max_op < header.op) max_op = header.op;
+                }
+
+                if (max_op != 0) {
+                    // Only committed ops are ever overwritten by a WAL wrap.
+                    assert(max_op - min_op < 2 * slot_count - config.pipelining_max);
```

Comment:
> I think this could fail, if for `journal_slot_count=10` as a backup we prepare `op=2` and then prepare `op=200`?

For `journal_slot_count=10`, I think the replica must not store (or ack) ops that far ahead of its current `op_checkpoint` (the last snapshotted op). Otherwise it will just have to discard+repair them anyway when it goes to try to commit & catch up.

That constraint what this assertion is verifying.